[inductor] Introduce CSEVariable type and use it to track if Triton variables are scalar

fdrocha opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88348
* __->__ #88347

This fixes https://github.com/pytorch/torchdynamo/issues/1515

To fix it, we need to keep track of whether a Triton variable is a scalar (so we can not use a mask when doing indirect loads through them). This requires a way of annotating variable names generated by CSE with properties.

So now CSE will use CSEVariable class to keep track of variables and let backends subclass it so they can annotate them with whatever information they want. TritonCSEVariable is such a subclass that track the `is_scalar` property.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire @lezcano

Commit:
[inductor] Track if variables are scalars

This fixes https://github.com/pytorch/torchdynamo/issues/1515

[ghstack-poisoned]

fdrocha commented:
@ngimel sorry, not sure what happened, had to close previous PR and do a new one. I copied your comment below:

> Something went wrong with ghstack, this pr is only formatting changes, and the real meat is in #88336 which is closed, > can you reopen in a more convenient shape? The general approach is fine, however I'd like the fix to also cover cases > like #1654 where (it looks) the problematic tensors are 1-element but not scalars. Let me know if you have trouble > reproing #1654
> IIRC, the #1654 problem is that triton allows mask can be broadcasted to tensor, but not the other way around (which is reasonable), but we generate 1-element tensor and 128 element mask.

I will look to address the above.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -277,7 +277,7 @@ def index_expr(expr, dtype):
     def masked(mask, body, other):
         code = BracesBuffer()
-        var = V.kernel.cse.newvar()
+        var = V.kernel.cse.newvar(is_scalar=False)
```
The arg name `is_scalar` is kind of confusing for cpp codegen. Variables are by default scalars in cpp and we are adding vectorization support but we manage the vector variable and scalar locally. Is it better named as `is_triton_scalar` and make it default to `False` to avoid changing cpp codegen? cc @EikanWang who is working on cpp vectorization support.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -336,7 +336,8 @@ def load(self, name: str, index: sympy.Expr):
             line = f"static_cast<float>({line})"
-        return self.cse.generate(self.loads, line)
+        # Not tracking whether results of loads are scalar
+        return self.cse.generate(self.loads, line, is_scalar_expr=False)
```
ditto.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -355,8 +356,12 @@ def store(self, name, index, value, mode=None):
-            self.loads, f"reduction {name} {cexpr(index)}", write=False
+            self.loads,
+            f"reduction {name} {cexpr(index)}",
+            is_scalar_expr=False,
```
ditto

ngimel commented:
@fdrocha is this ready to review? Does it handle #1654 or do you want to give up on handling #1654 for now? Also please address @jgong5's comments

fdrocha commented:
> @fdrocha is this ready to review? Does it handle #1654 or do you want to give up on handling #1654 for now? Also please address @jgong5's comments

No, please hold off. 
I've been thinking about how to best handle #1654, and it will probably require changing the approach here a bit.
When I do that, I will do it in away that addresses @jgon5's points, don't think it's worth doing in the present form.

Commit:
Update on "[inductor] Track if variables are scalars"


This fixes https://github.com/pytorch/torchdynamo/issues/1515

It keeps track of which variables in generated code are scalars, and does not include a mask when doing loads generated by these.

It uses variable names (e.g., tmp_scalar4) to keep track of which ones are scalars, which makes generated code a little more readable, but is a little hacky. Can try to make it better, but wanted to get some feedback on overall approach first.

This only really matters for triton backend, it would be nice to contain it to triton.py by didn't find a nice way of doing that.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx lezcano

[ghstack-poisoned]

fdrocha commented:
I have rewritten this to better separate the Triton specific aspects and make it a little more generic. This required introducing a new type `CSEVariable` to keep track of the `is_scalar` property, that could be extended for other properties if it is necessary. With the new approach `is_scalar` only appears on the triton backend. @jgong5 I think this addresses your concerns, please let me know.

@ngimel after thinking about it for a bit, it seems to me https://github.com/pytorch/torchdynamo/issues/1654 is related but different enough that it is better to handle it separately. I will try to put up a PR for it as well.
Otherwise, this is ready to review, so please have a look and let me know what you think when you have a chance.

ngimel commented on torch/_inductor/codegen/common.py:
```diff
@@ -430,9 +447,11 @@ def clone(self):
+    def generate(
+        self, buffer: IndentedBuffer, expr: typing.Union[str, CSEVariable], write=True
+    ) -> CSEVariable:
+        assert isinstance(expr, str) or isinstance(expr, CSEVariable), type(expr)
```
nit:
```
assert isinstance(expr, (str, CSEVariable))
```

ngimel reviewed (commented):
Please add a testcase

Commit:
Update on "[inductor] Introduce CSEVariable type and use it to track if Triton variables are scalar"


This fixes https://github.com/pytorch/torchdynamo/issues/1515

To fix it, we need to keep track of whether a Triton variable is a scalar (so we can not use a mask when doing indirect loads through them). This requires a way of annotating variable names generated by CSE with properties.

So now CSE will use CSEVariable class to keep track of variables and let backends subclass it so they can annotate them with whatever information they want. TritonCSEVariable is such a subclass that track the `is_scalar` property.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx lezcano

[ghstack-poisoned]

fdrocha commented:
> Please add a testcase

The next PR in this stack is a more comprehensive test for this. I added a simple more explicit test to this one just now.

jgong5 reviewed (approved):
LGTM. Thanks for the abstraction!

fdrocha commented:
@pytorchbot merge
